lib/json5/parser.ml

```diff
@@ -0,0 +1,84 @@
+open Let_syntax.Result
+
+let rec parse_list acc = function
+  | [] -> Error "List never ends"
+  | Lexer.CLOSE_BRACKET :: xs | COMMA :: CLOSE_BRACKET :: xs -> Ok (acc, xs)
+  | xs -> (
+      let* v, xs = parse xs in
+      match xs with
+      | [] -> Error "List was not closed"
+      | Lexer.CLOSE_BRACKET :: xs | COMMA :: CLOSE_BRACKET :: xs ->
+          Ok (v :: acc, xs)
+      | COMMA :: xs -> parse_list (v :: acc) xs
+      | x :: _ ->
+          let s =
+            Format.asprintf "Unexpected list token: %a" Lexer.pp_token x
+          in
+          Error s)
+
+and parse_assoc acc = function
+  | [] -> Error "Assoc never ends"
+  | Lexer.CLOSE_BRACE :: xs | COMMA :: CLOSE_BRACE :: xs -> Ok (acc, xs)
+  | STRING k :: COLON :: xs | IDENTIFIER_NAME k :: COLON :: xs -> (
+      let* v, xs = parse xs in
+      let item = (k, v) in
+      match xs with
+      | [] -> Error "Object was not closed"
+      | Lexer.CLOSE_BRACE :: xs | COMMA :: CLOSE_BRACE :: xs ->
+          Ok (item :: acc, xs)
+      | COMMA :: xs -> parse_assoc (item :: acc) xs
+      | x :: _ ->
+          let s =
+            Format.asprintf "Unexpected assoc list token: %a" Lexer.pp_token x
+          in
+          Error s)
+  | x :: _ ->
+      let s =
+        Format.asprintf "Unexpected assoc list token: %a" Lexer.pp_token x
+      in
+      Error s
+
+and parse = function
+  | [] -> Error "empty list of tokens"
+  | token :: xs -> (
+      match token with
+      | TRUE -> Ok (Ast.Bool true, xs)
+      | FALSE -> Ok (Bool false, xs)
+      | NULL -> Ok (Null, xs)
+      | INT v -> Ok (IntLit v, xs)
+      | FLOAT v -> Ok (FloatLit v, xs)
+      | INT_OR_FLOAT v -> Ok (FloatLit v, xs)
+      | STRING s -> Ok (StringLit s, xs)
+      | OPEN_BRACKET ->
+          let* l, xs = parse_list [] xs in
+          Ok (Ast.List (List.rev l), xs)
+      | OPEN_BRACE ->
+          let* a, xs = parse_assoc [] xs in
+          Ok (Ast.Assoc (List.rev a), xs)
+      | x ->
+          let s = Format.asprintf "Unexpected token: %a" Lexer.pp_token x in
+          Error s)
+
+let parse_from_lexbuf ?fname ?lnum lexbuffer =
+  let fname = Option.value fname ~default:"" in
+  let _ = Sedlexing.set_filename lexbuffer fname in
```

Comment:
This should probably be `;`-ed, instead of `_`-ed, to make sure this is `unit`. Same as below.